File: concepts/08_async_and_concurrency/05_shared_memory_atomics_and_synchronization/lua/test_02_lost_updates_and_predicates.lua
-- Common question: how should waiting use state predicates rather than timing guesses?
-- Inputs: an explicit readiness predicate, a producer coroutine, a consumer loop, and finite resumes.
-- Observations: predicate recheck, value publication, no sleep, and absent mutex/condition primitives.
-- polyglot-family: async_and_concurrency
-- polyglot-concept: shared_memory_atomics_and_synchronization
-- polyglot-related: languages/lua/language/08_coroutines_and_gc/test_059_yield_value_exchange.lua

local t = require("support.assertions")
local state = {ready = false}
local producer = coroutine.create(function()
    coroutine.yield("working")
    state.value = 42
    state.ready = true
end)

local resumes = 0
while not state.ready do
    local ok = coroutine.resume(producer)
    t.truth(ok)
    resumes = resumes + 1
    t.truth(resumes <= 2)
end
t.equal(resumes, 2)
t.equal(state.value, 42)
t.equal(rawget(_G, "Mutex"), nil)
t.equal(rawget(_G, "Condition"), nil)

-- Predicate loop 是当前宿主的 cooperative 协议，不伪装成线程 condition variable。
t.done()

File: concepts/08_async_and_concurrency/05_shared_memory_atomics_and_synchronization/ruby/test_02_lost_updates_and_predicates.rb
# 共同问题：如何可执行地展示 lost update，并用状态 predicate 而非 sleep 等待。
# 输入：同步 read/write 阶段、ConditionVariable、Mutex 和 ready flag；观察：确定性覆盖及 while predicate。
# polyglot-family: async_and_concurrency
# polyglot-concept: shared_memory_atomics_and_synchronization
# polyglot-related: languages/ruby/standard_library/11_concurrency_models/
# polyglot-related+: test_082_mutex_condition_variable_and_predicate.rb

require "assertions"

A = PolyglotAssertions

counter = 0
reads_complete = Queue.new
writes_allowed = Queue.new
workers = 2.times.map do
  Thread.new do
    observed = counter
    reads_complete << true
    writes_allowed.pop
    counter = observed + 1
  end
end
2.times { reads_complete.pop }
2.times { writes_allowed << true }
workers.each(&:join)
A.equal(1, counter)

mutex = Mutex.new
condition = ConditionVariable.new
ready = false
producer = Thread.new do
  mutex.synchronize do
    ready = true
    condition.signal
  end
end
mutex.synchronize { condition.wait(mutex) until ready }
producer.join
A.truth(ready)

A.done
